File: python/inverse_graph.py
from __future__ import annotations

# coding: utf-8
# inverse_graph.py
# Nodeを反転すべきかどうかを決めるグラフ

from itertools import count, product
from collections import defaultdict, Counter
from typing import List, Tuple, Optional, IO, Dict, Iterator
import random
import time

from graph import GraphBase, Node


#################### Type ####################

Edge = Tuple[Node, Node, int, int]
Value = Tuple[Node, int, int]


def is_consistent(graph: Dict[int, list[tuple[int, bool]]]) -> bool:
	visited: Dict[int, bool] = { }
	v0 = next(v for v in graph.keys())
	stk = [(v0, False)]		# v0は反転しないとする
	while stk:
		v, b = stk.pop()
		for v1, b1_edge in graph[v]:
			b1 = b ^ b1_edge
			if v1 in visited:
				b1_prev = visited[v1]
				if b1 != b1_prev:
					return False
			else:
				visited[v1] = b1
				stk.append((v1, b1))
	return True
	
def invs(graph: Dict[Node, list[tuple[Node, bool]]]) -> Dict[Node, bool]:
	visited: Dict[Node, bool] = { }
	v0 = min(v for v in graph.keys())
	stk = [(v0, False)]		# v0は反転しないとする
	while stk:
		v, b = stk.pop()
		for v1, b1_edge in graph[v]:
			b1 = b ^ b1_edge
			if v1 not in visited:
				visited[v1] = b1
				stk.append((v1, b1))
	return visited


#################### InverseGraph ####################

class InverseGraph(dict, GraphBase):
	def __init__(self, *args, **kwargs):
		dict.__init__(self, *args, **kwargs)
		GraphBase.__init__(self)
	
	def __setitem__(self, key: Node, value: list[Value]):
		super().__setitem__(key, value)
	
	def copy(self) -> InverseGraph:
		graph = InverseGraph()
		for v, vs in self.items():
			graph[v] = vs[:]
		return graph
	
	##### virtual methods for GraphBase #####
	def generate_nodes(self) -> Iterator[Node]:
		for v in self.keys():
			yield v
	
	def neighbors(self, v0: Node) -> list[Node]:
		value = self.get(v0, [])
		return [ v for v, n1, n2 in value ]
	
	##### non-virtual methods #####
	def optimize_inversions(self) -> tuple[bool, ...]:
		subgraphs = self.divide_graph_into_connected()
		bs = [False] * len(self)
		for subg in subgraphs:
			dic_bs = subg.optimize_inversions_connected()
			for i, b in dic_bs.items():
				bs[i] = b
		return tuple(bs)
	
	def divide_graph_into_connected(self) -> list[InverseGraph]:
		vss = self.divide_nodes_into_connected()
		subgraphs: list[InverseGraph] = []
		for vs in vss:
			subg = InverseGraph()
			for v in vs:
				subg[v] = self.get(v, [])
			subgraphs.append(subg)
		return subgraphs
	
	def is_leaf(self, v: Node) -> bool:
		return v in self and len(self[v]) == 1
	
	def remove_directed_edge(self, v1: Node, v2: Node):
		neighs1 = self.neighbors(v1)
		if len(neighs1) == 1:
			del self[v1]
		else:
			vs2 = self[v1]
			for i, (v, n1, n2) in enumerate(vs2):
				if v == v2:
					vs2.pop(i)
					break
			else:
				raise Exception('error in InverseGraph.remove_directed_edge')
	
	def remove_edge(self, v1: Node, v2: Node):
		self.remove_directed_edge(v1, v2)
		self.remove_directed_edge(v2, v1)
	
	def remove_acyclic_nodes(self) -> tuple[InverseGraph, list[Node]]:
		graph = self.copy()
		removed_nodes: list[Node] = []
		while True:
			prev_num_removed_nodes = len(removed_nodes)
			vs = list(graph.keys())
			for v in vs:
				if graph.is_leaf(v):
					neighs = graph.neighbors(v)
					v1 = neighs[0]
					graph.remove_edge(v, v1)
					removed_nodes.append(v)
			
			if len(removed_nodes) == prev_num_removed_nodes:
				break
		return (graph, removed_nodes)
	
	# assume connected
	def is_tree(self) -> bool:
		num_edges = sum(len(vs) for vs in self.values())
		return num_edges == (len(self) - 1) * 2
	
	def decide_inversions_tree(self) -> Dict[Node, bool]:
		dic_bs = { }
		v0 = next(v for v in sorted(self.keys()))
		visited = set([v0])
		stk = [(v0, False)]
		dic_bs[v0] = False
		while stk:
			(v, b) = stk.pop()
			for v1, n1, n2 in self[v]:
				b1 = b ^ (n1 < n2)
				if v1 not in visited:
					dic_bs[v1] = b1
					stk.append((v1, b1))
					visited.add(v1)
		return dic_bs
	
	# Assume that self is a connected component
	def optimize_inversions_connected_core(self) -> Dict[Node, bool]:
		if len(self) <= 20:
			return self.search_all()
		
		# Finalize the edge from the one with the largest true/false bias
		# If there are no inconsistencies by the time the graphs are all connected, it's OK.
		dic_bs = self.connect_biased_edges()
		if dic_bs:
			return dic_bs
		
		# If that doesn't work, use random.
		# I really want to use SA.
		return self.search_randomly()
	
	def add_removed_nodes(self, dic_bs: Dict[Node, bool],
								removed_nodes: list[Node]):
		for v1 in reversed(removed_nodes):
			for v2, n1, n2 in self[v1]:
				if v2 in dic_bs:
					b2 = dic_bs[v2]
					b1 = b2 ^ (n1 < n2)
					dic_bs[v1] = b1
					break
	
	# Assume that self is a connected component
	def optimize_inversions_connected(self) -> Dict[Node, bool]:
		if self.is_tree():
			return self.decide_inversions_tree()
		
		graph, removed_nodes = self.remove_acyclic_nodes()
		dic_bs = graph.optimize_inversions_connected_core()
		self.add_removed_nodes(dic_bs, removed_nodes)
		return dic_bs
	
	# Brute force
	def search_all(self) -> Dict[Node, bool]:
		if len(self) == 1:
			return { list(self.keys())[0]: False }
		
		bs = min(product((False, True), repeat=len(self)-1),
										key=lambda bs: self.score(bs))
		vs = sorted(self.keys())
		return dict(zip(vs, (False,) + bs))
	
	def search_randomly(self) -> Dict[Node, bool]:
		bs = min((tuple(random.choices((False, True), k=len(self)-1))
												for _ in range(2**19)),
												key=lambda bs: self.score(bs))
		vs = sorted(self.keys())
		return dict(zip(vs, (False,) + bs))
	
	def generate_biased_edges(self) -> Iterator[Edge]:
		for i, v in self.items():
			for j, n1, n2 in v:
				if i > j:
					continue
				yield (i, j, n1, n2)
	
	def score(self, bs: tuple[bool, ...]) -> int:
		dic: Dict[int, int] = dict(zip(sorted(self.keys()), count()))
		num_wrong: int = 0
		for v1, v2, n1, n2 in self.generate_biased_edges():
			i = dic[v1]
			j = dic[v2]
			b1 = False if i == 0 else bs[i-1]
			b2 = bs[j-1]
			if b1 == b2:
				num_wrong += n2
			else:
				num_wrong += n1
		return num_wrong
	
	def connect_biased_edges(self) -> Dict[Node, bool]:
		def ratio(v: Edge) -> float:
			i, j, n1, n2 = v
			return min((n1 + 1) / (n1 + n2 + 2), (n2 + 1) / (n1 + n2 + 2))
		
		def join(graph1: Dict[int, list[tuple[int, bool]]],
							graph2: Dict[int, list[tuple[int, bool]]]):
			for v1, vs in graph2.items():
				graph1[v1] = vs
		
		# これってminimum spanning treeなのでは
		sorted_edges = sorted(((ratio(e), e)
								for e in self.generate_biased_edges()))
		subgraphs: list[BoolGraph] = []
		for v in self.keys():
			graph = BoolGraph()
			graph[v] = []
			subgraphs.append(graph)
		
		for r, (i, j, n1, n2) in sorted_edges:
			k1 = next(k for k, g in enumerate(subgraphs) if i in g.keys())
			k2 = next(k for k, g in enumerate(subgraphs) if j in g.keys())
			if k1 == k2:
				g1 = subgraphs[k1]
				g1[i].append((j, n1 < n2))
				g1[j].append((i, n1 < n2))
				if not is_consistent(g1):
					return { }
			else:
				subg1 = subgraphs[k1]
				subg2 = subgraphs.pop(k2)
				join(subg1, subg2)
				subg1[i].append((j, n1 < n2))
				subg1[j].append((i, n1 < n2))
				if len(subgraphs) == 1:
					return invs(subgraphs[0])
		else:
			return { }	# not come here
	
	@staticmethod
	def convert(graph: list[list[tuple[int, int, int]]]) -> InverseGraph:
		return InverseGraph(dict(enumerate(graph)))


#################### BoolGraph ####################

class BoolGraph(dict, GraphBase):
	def __init__(self, *args, **kwargs):
		dict.__init__(self, *args, **kwargs)
		GraphBase.__init__(self)
	
	def __setitem__(self, key: Node, value: list[Value]):
		super().__setitem__(key, value)
	
	##### virtual methods for GraphBase #####
	def generate_nodes(self) -> Iterator[Node]:
		for v in self.keys():
			yield v
	
	def neighbors(self, v0: Node) -> list[Node]:
		value = self.get(v0, [])
		return [ v for v, b in value ]
	
	##### non-virtual methods #####
	def join(self, graph: BoolGraph):
		for u, vs in graph.items():
			self.__setitem__(u, vs)


__all__ = ['InverseGraph']

File: src/inverse_graph.cpp
#include <set>
#include <stack>
#include <algorithm>
#include <cassert>
#include <random>
#include "../include/inverse_graph.h"

using namespace std;


//////////////////// InverseGraph ////////////////////

InverseGraph *InverseGraph::copy() const {
	auto	*g = new InverseGraph();
	for(auto p = begin(); p != end(); ++p)
		(*g)[p->first] = p->second;
	return g;
}

vector<InverseGraph::Node> InverseGraph::collect_nodes() const {
	vector<Node>	nodes;
	for(auto p = begin(); p != end(); ++p)
		nodes.push_back(p->first);
	return nodes;
}

vector<InverseGraph::Node> InverseGraph::neighbors(Node v0) const {
	auto	p = find(v0);
	if(p == end())
		return vector<Node>();
	else {
		vector<Node>	neighs;
		for(auto q = p->second.begin(); q != p->second.end(); ++q)
			neighs.push_back(get<0>(*q));
		return neighs;
	}
}

vector<InverseGraph::Node> InverseGraph::sort_nodes() const {
	vector<Node>	vs;
	for(auto p = begin(); p != end(); ++p)
		vs.push_back(p->first);
	return vs;
}

map<InverseGraph::Node, size_t> InverseGraph::calc_dic() const {
	map<Node, size_t>	dic;
	size_t	i = 0;
	for(auto p = begin(); p != end(); ++i, ++p)
		dic[p->first] = i;
	return dic;
}

vector<bool> InverseGraph::optimize_inversions() const {
	const auto	subgraphs = this->divide_into_connected();
	vector<bool>	bs(this->size(), false);
	for(auto p = subgraphs.begin(); p != subgraphs.end(); ++p) {
		const InverseGraph&	subg = *p;
		const auto	dic_bs = subg.optimize_inversions_connected();
		for(auto q = dic_bs.begin(); q != dic_bs.end(); ++q)
			bs[q->first] = q->second;
	}
	return bs;
}

vector<InverseGraph> InverseGraph::divide_into_connected() const {
	const vector<vector<Node>>	nss = divide_nodes_into_connected();
	vector<InverseGraph>	graphs(nss.size());
	for(size_t i = 0; i < nss.size(); ++i) {
		for(auto p = nss[i].begin(); p != nss[i].end(); ++p) {
			const Node	v = *p;
			auto	q = find(v);
			graphs[i][v] = q->second;
		}
	}
	return graphs;
}

bool InverseGraph::is_leaf(Node v) const {
	auto	p = find(v);
	return p != end() && p->second.size() == 1;
}

void InverseGraph::remove_directed_edge(Node v1, Node v2) {
	const auto	neighs1 = this->neighbors(v1);
	if(neighs1.size() == 1)
		this->erase(v1);
	else {
		auto&	vs2 = (*this)[v1];
		for(auto p = vs2.begin(); p != vs2.end(); ++p) {
			if(get<0>(*p) == v2) {
				vs2.erase(p, p+1);
				return;
			}
		}
		throw("error in InverseGraph::remove_directed_edge");
	}
}

void InverseGraph::remove_edge(Node v1, Node v2) {
	this->remove_directed_edge(v1, v2);
	this->remove_directed_edge(v2, v1);
}

pair<InverseGraph *, vector<InverseGraph::Node>>
					InverseGraph::remove_acyclic_nodes() const {
	auto	*graph = this->copy();
	vector<Node>	removed_nodes;
	while(true) {
		const size_t	prev_num_removed_nodes = removed_nodes.size();
		const auto	vs = graph->collect_nodes();
		for(auto p = vs.begin(); p != vs.end(); ++p) {
			const Node	v = *p;
			if(graph->is_leaf(v)) {
				const auto	neighs = graph->neighbors(v);
				const Node	v1 = neighs[0];
				graph->remove_edge(v, v1);
				removed_nodes.push_back(v);
			}
		}
		if(removed_nodes.size() == prev_num_removed_nodes)
			break;
	}
	return make_pair(graph, removed_nodes);
}

// assume connected
bool InverseGraph::is_tree() const {
	size_t	num_edges = 0;
	for(auto p = begin(); p != end(); ++p)
		num_edges += p->second.size();
	return num_edges == (size() - 1) * 2;
}

map<InverseGraph::Node, bool> InverseGraph::decide_inversions_tree() const {
	map<InverseGraph::Node, bool>	dic_bs;
	const Node	v0 = begin()->first;
	set<Node>	visited;
	visited.insert(v0);
	vector<pair<Node, bool>>	stk(1, make_pair(v0, false));
	dic_bs[v0] = false;
	while(!stk.empty()) {
		const auto	_p = stk.back();
		stk.pop_back();
		const Node	v = _p.first;
		const bool	b = _p.second;
		auto	p = find(v);
		if(p == end())
			continue;
		for(auto q = p->second.begin(); q != p->second.end(); ++q) {
			const Node	v1 = get<0>(*q);
			const int	n1 = get<1>(*q);
			const int	n2 = get<2>(*q);
			const bool	b1 = b ^ (n1 < n2);
			if(visited.find(v1) == visited.end()) {
				dic_bs[v1] = b1;
				stk.push_back(make_pair(v1, b1));
				visited.insert(v1);
			}
		}
	}
	return dic_bs;
}

map<InverseGraph::Node, bool>
				InverseGraph::optimize_inversions_connected_core() const {
	if(this->size() <= 20)
		return this->search_all();
	
	// Determine edge from those with large true/false bias
	// It is OK if there is no contradiction until all the graphs are connected.
	const auto	dic_bs = this->connect_biased_edges();
	if(!dic_bs.empty())
		return dic_bs;
	
	// decide to flip randomly if contradictory
	// I really want to use SA
	return this->search_randomly();
}

void InverseGraph::add_removed_nodes(map<InverseGraph::Node, bool>& dic_bs,
									const vector<Node>& removed_nodes) const {
	for(auto p = removed_nodes.rbegin(); p != removed_nodes.rend(); ++p) {
		const Node	v1 = *p;
		const auto	vs = find(v1)->second;
		for(auto q = vs.begin(); q != vs.end(); ++q) {
			const Node	v2 = get<0>(*q);
			const int	n1 = get<1>(*q);
			const int	n2 = get<2>(*q);
			auto	r = dic_bs.find(v2);
			if(r != dic_bs.end()) {
				const bool	b2 = r->second;
				const bool	b1 = b2 ^ (n1 < n2);
				dic_bs.insert(make_pair(v1, b1));
				break;
			}
		}
	}
}

// graph must be connected
map<InverseGraph::Node, bool>
					InverseGraph::optimize_inversions_connected() const {
	if(this->is_tree())
		return this->decide_inversions_tree();
	
	const auto	p = this->remove_acyclic_nodes();
	auto	*graph = p.first;
	auto	removed_nodes = p.second;
	auto	dic_bs = graph->optimize_inversions_connected_core();
	this->add_removed_nodes(dic_bs, removed_nodes);
	delete graph;
	return dic_bs;
}

// brute force
map<InverseGraph::Node, bool> InverseGraph::search_all() const {
	const size_t	N = this->size();
	int	min_score = 100000000;
	vector<bool>	min_invs;
	for(int i = 0; i < (1 << (N-1)); ++i) {
		// Instead of converting it to bools,
		// it's faster to keep it as an integer.
		vector<bool>	invs(N-1);
		for(size_t j = 0; j < N - 1; ++j)
			invs[j] = ((i >> j) & 1) == 1;
		
		const int	score = this->calc_match_score(invs);
		if(score < min_score) {
			min_score = score;
			min_invs = invs;
		}
	}
	
	map<Node, bool>	dic;
	const vector<Node>	vs = sort_nodes();
	dic[vs[0]] = false;
	for(size_t i = 1; i < vs.size(); ++i)
		dic[vs[i]] = min_invs[i-1];
	return dic;
}

int InverseGraph::choice_seed() const {
	int	s = 0;
	int	D = 1000000000;
	for(auto p = begin(); p != end(); ++p) {
		for(auto q = p->second.begin(); q != p->second.end(); ++q)
			s = (s + (int)(get<0>(*q))) + D;
	}
	return s;
}

std::map<InverseGraph::Node, bool> InverseGraph::search_randomly() const {
	vector<bool>	min_bs;
	int				min_score = 1000000000;
	std::mt19937 generator(this->choice_seed());
	std::bernoulli_distribution	distribution(0.5);
	for(int i = 0; i < 524288; ++i) {	// 2^19
		vector<bool>	bs(this->size() - 1);
		for(size_t k = 0; k < this->size() - 1; ++k)
			bs.push_back(distribution(generator));
		const int	cur_score = this->calc_match_score(bs);
		if(min_bs.empty() || cur_score < min_score) {
			min_bs = bs;
			min_score = cur_score;
		}
	}
	
	map<Node, bool>	dic;
	const vector<Node>	vs = sort_nodes();
	dic[vs[0]] = false;
	for(size_t i = 1; i < vs.size(); ++i)
		dic[vs[i]] = min_bs[i-1];
	return dic;
}

vector<InverseGraph::Edge> InverseGraph::extract_edges() const {
	vector<Edge>	edges;
	for(auto p = begin(); p != end(); ++p) {
		const Node&	v1 = p->first;
		for(auto q = p->second.begin(); q != p->second.end(); ++q) {
			const Node&	v2 = get<0>(*q);
			const int&	n1 = get<1>(*q);
			const int&	n2 = get<2>(*q);
			if(v1 < v2)
				edges.push_back(Edge(v1, v2, n1, n2));
		}
	}
	return edges;
}

int InverseGraph::calc_match_score(const vector<bool>& bs) const {
	int	num_wrong = 0;
	const auto	edges = this->extract_edges();
	const auto	dic_vs = calc_dic();
	for(auto p = edges.begin(); p != edges.end(); ++p) {
		auto	p1 = dic_vs.find(get<0>(*p));
		auto	p2 = dic_vs.find(get<1>(*p));
		const size_t	i = p1->second;
		const size_t	j = p2->second;
		const bool	b1 = i == 0 ? false : bs[i-1];
		const bool	b2 = bs[j-1];
		if(b1 == b2)
			num_wrong += get<3>(*p);
		else
			num_wrong += get<2>(*p);
	}
	return num_wrong;
}

vector<pair<double, InverseGraph::Edge>> InverseGraph::sort_edges() const {
	const vector<Edge>	edges = this->extract_edges();
	vector<pair<double, Edge>>	sorted_edges;
	for(auto p = edges.begin(); p != edges.end(); ++p) {
		const int	n1 = get<2>(*p);
		const int	n2 = get<3>(*p);
		const double	r1 = (n1 + 1) / (double)(n1 + n2 + 2);
		const double	r2 = (n2 + 1) / (double)(n1 + n2 + 2);
		const double	ratio = std::min(r1, r2);
		sorted_edges.push_back(make_pair(ratio, *p));
	}
	std::sort(sorted_edges.begin(), sorted_edges.end());
	return sorted_edges;
}

map<InverseGraph::Node, bool> InverseGraph::connect_biased_edges() const {
	const vector<pair<double, Edge>>	sorted_edges = this->sort_edges();
	// [ { node: [(index, inv?)] } ]
	vector<BoolGraph>	subgraphs;
	for(auto p = begin(); p != end(); ++p) {
		BoolGraph	graph;
		graph[p->first];
		subgraphs.push_back(graph);
	}
	
	// Add from highly biased edge
	for(auto p = sorted_edges.begin(); p != sorted_edges.end(); ++p) {
		const size_t	i = get<0>(p->second);
		const size_t	j = get<1>(p->second);
		const int		n1 = get<2>(p->second);
		const int		n2 = get<3>(p->second);
		int	k1;
		for(k1 = 0; ; ++k1) {
			const auto&	g1 = subgraphs[k1];
			if(g1.find(i) != g1.end())
				break;
		}
		int	k2;
		for(k2 = 0; ; ++k2) {
			const auto&	g2 = subgraphs[k2];
			if(g2.find(j) != g2.end())
				break;
		}
		if(k1 == k2) {
			auto&	subg1 = subgraphs[k1];
			subg1[i].push_back(make_pair(j, n1 < n2));
			subg1[j].push_back(make_pair(i, n1 < n2));
			if(!InverseGraph::is_consistent(subgraphs[k1]))
				return map<size_t, bool>();
		}
		else {
			auto&	subg1 = subgraphs[k1];
			auto&	subg2 = subgraphs[k2];
			subg1.join(subg2);
			subg1[i].push_back(make_pair(j, n1 < n2));
			subg1[j].push_back(make_pair(i, n1 < n2));
			subgraphs.erase(subgraphs.begin() + k2);
			if(subgraphs.size() == 1)
				return InverseGraph::invs(subgraphs[0]);
		}
	}
	return map<size_t, bool>();		// not coming here
}

bool InverseGraph::is_consistent(
						const map<Node, vector<pair<Node, bool>>>& graph) {
	map<Node, bool>	visited;
	Node	v0 = graph.begin()->first;
	stack<pair<Node, bool>>	stk;
	stk.push(make_pair(v0, false));
	while(!stk.empty()) {
		const Node	v = stk.top().first;
		const bool	b = stk.top().second;
		stk.pop();
		auto	iter = graph.find(v);
		for(auto p = iter->second.begin(); p != iter->second.end(); ++p) {
			const Node	v1 = p->first;
			const bool	b1_edge = p->second;
			const bool	b1 = b ^ b1_edge;
			auto	q = visited.find(v1);
			if(q != visited.end()) {
				const bool	b1_prev = q->second;
				if(b1 != b1_prev)
					return false;
			}
			else {
				visited.insert(make_pair(v1, b1));
				stk.push(make_pair(v1, b1));
			}
		}
	}
	return true;
}

map<InverseGraph::Node, bool> InverseGraph::invs(
							const map<Node, vector<pair<Node, bool>>>& graph) {
	map<Node, bool>	visited;
	const Node	v0 = graph.begin()->first;
	stack<pair<Node, bool>>	stk;
	stk.push(make_pair(v0, false));		// assume that v0 is not inverted
	while(!stk.empty()) {
		const Node	v = stk.top().first;
		const bool	b = stk.top().second;
		stk.pop();
		auto	iter = graph.find(v);
		for(auto p = iter->second.begin(); p != iter->second.end(); ++p) {
			const Node	v1 = p->first;
			const bool	b1_edge = p->second;
			const bool	b1 = b ^ b1_edge;
			if(visited.find(v1) == visited.end()) {
				visited.insert(make_pair(v1, b1));
				stk.push(make_pair(v1, b1));
			}
		}
	}
	return visited;
}

ostream& operator <<(ostream& os, const InverseGraph& graph) {
	for(auto p = graph.begin(); p != graph.end(); ++p) {
		os << p->first << " : ";
		for(auto q = p->second.begin(); q != p->second.end(); ++q)
			os << "(" << get<0>(*q) << ", " << get<1>(*q)
							<< ", " << get<2>(*q) << ") ";
		os << "\n";
	}
	return os;
}


//////////////////// BoolGraph ////////////////////

vector<BoolGraph::Node> BoolGraph::collect_nodes() const {
	vector<Node>	nodes;
	for(auto p = begin(); p != end(); ++p)
		nodes.push_back(p->first);
	return nodes;
}

vector<BoolGraph::Node> BoolGraph::neighbors(Node v0) const {
	auto	p = find(v0);
	if(p == end())
		return vector<Node>();
	else {
		vector<Node>	neighs;
		for(auto q = p->second.begin(); q != p->second.end(); ++q)
			neighs.push_back(get<0>(*q));
		return neighs;
	}
}

void BoolGraph::join(const BoolGraph& graph) {
	for(auto p = graph.begin(); p != graph.end(); ++p)
		insert(*p);
}
